File: php/import_professionals.php
<?php
include "settings.php";

function get_professionals($base_url, $token) {
    $url = $base_url . 'p/items';
    $result = json_decode(get($url, $token));
    return $result->{'@items'};
}

function get_professional($base_url, $token, $id=False, $email=False) {
    if ((!$id and !$email) or ($id and $email)) {
        exit("Please search for id or email.");
    }
    $professionals = get_professionals($base_url, $token);
    foreach ($professionals as $professional) {
        if ($id and $professional->id == $id) {
            return $professional;
        } else if ($email and $professional->email == $email) {
            return $professional;
        }
    }
    return False;
}

function add_professional(
        $base_url, $token, $first_name, $infix, $last_name, $id, $email) {
    $url = $base_url . 'p';

    $payload = array(
        "active" => true,
        "first_name" => $first_name,
        "infix" => $infix,
        "last_name" => $last_name,
        "id" => $id,
        "email" => $email
    );
    $result = json_decode(post($url, $payload, $token));

    if (property_exists($result, 'code') && $result->code != 201) {
        print $payload['email'];
        print $payload['id'];
        print "\n";
        foreach ($result->extra as $err) {
            print $err->message;
            print "\n";
        }
    } else {
        return get_professional($base_url, $token, $id);
    }
}

function update_professional(
        $url, $token, $first_name, $infix, $last_name, $id, $email) {

    $payload = array(
        "first_name" => $first_name,
        "infix" => $infix,
        "last_name" => $last_name,
        "id" => $id,
        "email" => $email
    );
    patch($url, $payload, $token);
}

function add_random_professional($base_url, $token) {
    $first_name = random_first_name();
    $infix = random_infix();
    $last_name = random_last_name();
    return add_professional(
        $base_url,
        $token,
        $first_name,
        $infix,
        $last_name,
        random_id(),
        random_email($first_name, $infix, $last_name)
    );
}

$ROLES = array(
    'secretary' => 'ith.Secretary',
    'therapist' => 'ith.Therapist',
    'supervisor' => 'ith.Supervisor',
    'application manager' => 'ith.ApplicationManager',
    'analyst' => 'ith.Analyst'
);

function set_role($url, $token, $role) {
    global $ROLES;
    $url = $url . "/roles";
    $role = trim(strtolower($role));
    $payload = array(
        "roles" => array($ROLES[strtolower($role)])
    );

    return json_decode(patch($url, $payload, $token));
}

function import_professionals ($base_url, $token, $filepath) {
    $file = fopen($filepath, 'r');
    $header_seen = false;
    while (($line = fgetcsv($file)) !== FALSE) {
        if (!$header_seen) {
            $header_seen = true;
            continue;
        }
        $email = strtolower($line[0]);
        $id = $line[1];
        $first_name = $line[2];
        $infix = $line[3];
        $last_name = $line[4];
        $role = $line[5];

        $professional = get_professional($base_url, $token, $id);
        if($professional) {
            update_professional(
                $professional->{'@url'},
                $token,
                $first_name,
                $infix,
                $last_name,
                $id,
                $email);
            set_role($professional->{'@url'}, $token, $role);
            echo "Updated professional $email.\n";
        } else {
            $professional = add_professional(
                $base_url,
                $token,
                $first_name,
                $infix,
                $last_name,
                $id,
                $email);
            if ($professional) {
                set_role($professional->{'@url'}, $token, $role);
                echo "Added $email.\n";
            }
        }
    }
    fclose($file);
}

import_professionals($config_url, $config_token, $config_path);
# $professional_url = add_random_professional($config_url, $config_token);
# $result = set_role($professional_url, $config_token, "therapist");

File: python/import_professionals.py
#!/usr/bin/env python
"""This script uses the Minddistrict API in order to add
professionals from an CSV file.
"""
import csv
import util
import config

ROLE_MAP = {
    'secretary': 'ith.Secretary',
    'therapist': 'ith.Therapist',
    'supervisor': 'ith.Supervisor',
    'application manager': 'ith.ApplicationManager',
    'analyst': 'ith.Analyst'
}


def get_professionals_to_import(path):
    """Read the CSV file returning professional information as Python
    dictionaries.
    """
    professionals = []
    with open(path, 'r') as csvfile:
        reader = csv.reader(csvfile)
        reader.next()  # Skip the first "header" row.
        for row in reader:
            professionals.append({
                'email': row[0].strip(),
                'id': row[1].strip(),
                'first_name': unicode(row[2], "UTF-8").strip(),
                'infix': unicode(row[3], "UTF-8").strip(),
                'last_name': unicode(row[4], "UTF-8").strip(),
                'role': ROLE_MAP[row[5].strip().lower()],
            })
    return professionals


class Professional(object):

    def __init__(self, email, id, first_name, infix, last_name):
        self.professionals_url = config.url + '/p'
        self.email = email
        self.id = id
        self.first_name = first_name
        self.infix = infix
        self.last_name = last_name

    def create(self):
        print 'Create', self.email
        # TODO Raise error if problem.
        result = util.query_api(
            url=self.professionals_url,
            data={
                "email": self.email,
                "id": self.id,
                "first_name": self.first_name,
                "infix": self.infix,
                "last_name": self.last_name,
                "active": True},
            token=config.token,
            method='POST')
        self.url = result['@url']

    def set_role(self, role):
        url = self.url + '/roles'
        return util.query_api(
            url=url,
            data={"roles": [role]},
            token=config.token,
            method='PATCH')


def add_professionals(professionals, url, token):
    for professional_info in professionals:
        prof = Professional(
            professional_info['email'],
            professional_info['id'],
            professional_info['first_name'],
            professional_info['infix'],
            professional_info['last_name']
        )
        prof.create()
        prof.set_role(professional_info['role'])


if __name__ == '__main__':
    add_professionals(
        get_professionals_to_import(config.path),
        config.url,
        config.token)
    print "done"
